Yosemite/YosemiteTests/Storage/StorageManagerConcurrencyTests.swift

```diff
@@ -0,0 +1,270 @@
+        [firstDerivedStorage, secondDerivedStorage].forEach { derivedStorage in
+            derivedStorage.perform {
+                self.upsert(orderStatus: orderStatus, using: derivedStorage)
+            }
+        }
+
+        [firstDerivedStorage, secondDerivedStorage].forEach { derivedStorage in
+            waitForExpectation { exp in
+                storageManager.saveDerivedType(derivedStorage: derivedStorage) {
+                    exp.fulfill()
+                }
+            }
+        }
+
+        // Then
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.OrderStatus.self), 2)
+    }
+
+    /// Simulate what will happen if both the `perform()` and `saveDerivedType` happens at the
+    /// same time. We end up with the same record inserted twice.
+    ///
+    func testWhenConcurrentlySavingItCanAllowSavingOfDuplicates() {
+        // Given
+        let firstDerivedStorage = storageManager.newDerivedStorage()
+        let secondDerivedStorage = storageManager.newDerivedStorage()
+
+        let orderStatus = Networking.OrderStatus(name: "In Space", siteID: 1_998, slug: "in-space", total: 9)
+
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.OrderStatus.self), 0)
+
+        // When
+        let exp = expectation(description: "concurrent-saving")
+        exp.expectedFulfillmentCount = 2
+
+        [firstDerivedStorage, secondDerivedStorage].forEach { derivedStorage in
+            derivedStorage.perform {
+                self.upsert(orderStatus: orderStatus, using: derivedStorage)
+            }
+
+            self.storageManager.saveDerivedType(derivedStorage: derivedStorage) {
+                exp.fulfill()
+            }
+        }
+
+        wait(for: [exp], timeout: Constants.expectationTimeout)
+
+        // Then
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.OrderStatus.self), 2)
+    }
+
+    /// Simulate what will happen if `perform()` and `saveDerivedType` called under a single
+    /// `perform()` block happens at the same time on two `StorageTypes`. We end up with the same
+    /// record inserted twice.
+    ///
+    func testWhenConcurrentlySavingUsingASinglePerformBlockItCanAllowSavingOfDuplicates() {
+        // Given
+        let firstDerivedStorage = storageManager.newDerivedStorage()
+        let secondDerivedStorage = storageManager.newDerivedStorage()
+
+        let orderStatus = Networking.OrderStatus(name: "In Space", siteID: 1_998, slug: "in-space", total: 9)
+
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.OrderStatus.self), 0)
+
+        // When
+        let exp = expectation(description: "concurrent-saving")
+        exp.expectedFulfillmentCount = 2
+
+        [firstDerivedStorage, secondDerivedStorage].forEach { derivedStorage in
+            derivedStorage.perform {
+                self.upsert(orderStatus: orderStatus, using: derivedStorage)
+
+                self.storageManager.saveDerivedType(derivedStorage: derivedStorage) {
```

Comment:
Here the sequence it's correct: you call `saveDerivedType` immediately after the upsert, and both are inside the perform block.

Reply:
In [this discussion](https://github.com/woocommerce/woocommerce-ios/pull/2316/files#r429192587), I think we've discovered why we were getting a zero result. But yeah, this one here is also a valid pattern because the `perform()` will just be queued and everything should still happen serially.